framework/src/utils/PetscDMMoose.C

```diff
@@ -617,23 +619,54 @@ DMMooseGetEmbedding_Private(DM dm, IS * embedding)
               locator = displaced_problem->geomSearchData()._penetration_locators[it.first];
             }
             else
-            locator = dmm->_nl->_fe_problem.geomSearchData()._penetration_locators[it.first];
-
-            std::vector<dof_id_type> & slave_nodes = locator->_nearest_node._slave_nodes;
-            for (dof_id_type i = 0; i < slave_nodes.size(); ++i)
+              locator = dmm->_nl->_fe_problem.geomSearchData()._penetration_locators[it.first];
+            // penetration locator
+            std::map<dof_id_type, PenetrationInfo *>::iterator
+            lit  = locator->_penetration_info.begin(),
+            lend = locator->_penetration_info.end();
+            for (; lit!=lend; ++lit)
             {
-              if (locator->_has_penetrated.find(slave_nodes[i]) == locator->_has_penetrated.end())
-                continue;
-
-              Node & slave_node = dmm->_nl->sys().get_mesh().node_ref(slave_nodes[i]);
-              dof_id_type dof = slave_node.dof_number(dmm->_nl->sys().number(), v, 0);
-
-              // might want to use variable_first/last_local_dof instead
-              if (dof >= dofmap.first_dof() && dof < dofmap.end_dof())
-                indices.insert(dof);
-            }
-          }
-        }
+              const dof_id_type slave_node_num = lit->first;
+              PenetrationInfo * pinfo = lit->second;
+              std::map<dof_id_type, std::vector<dof_id_type> > & node_to_elem_map =
+              dmm->_nl->_fe_problem.mesh().nodeToElemMap();
+              if (pinfo && pinfo->isCaptured())
+              {
+                Node & slave_node = dmm->_nl->sys().get_mesh().node_ref(slave_node_num);
+                dof_id_type dof = slave_node.dof_number(dmm->_nl->sys().number(), v, 0);
+                // might want to use variable_first/last_local_dof instead
+                if (dof >= dofmap.first_dof() && dof < dofmap.end_dof())
+                  indices.insert(dof);
+                else
+                  cached_indices.insert(dof); // cache nonlocal indices
+                // indices of slave elements
+                std::vector<dof_id_type> evindices;
+                for (const auto & elem_num : node_to_elem_map[slave_node_num])
+                {
+                  Elem & slave_elem = dmm->_nl->sys().get_mesh().elem_ref(elem_num);
+                  // Get the degree of freedom indices for the given variable off the current element.
+                  evindices.clear();
```

Comment:
You should definitely size this vector to avoid several reallocations in the loop below.